03b_ipyautoui.ipynb

```diff
@@ -10,20 +10,749 @@
+    "    # fields have values that are valid.\n",
+    "    @model_validator(mode=\"after\")\n",
+    "    def limit_polynomial_order(self):\n",
+    "        \n",
+    "        if self.polynomial_order > self.window_size - 1:\n",
+    "            # Handle a bad polynomial order or window size\n",
+    "            raise ValueError(\"Polynomial order must be smaller than window size\")\n",
+    "            \n",
+    "        # If we got this far the polynomial order is consistent with the window size\n",
+    "        # so return self. Failing to return self will end up causing an error.\n",
+    "        return self"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ui_final = AutoUi(DataSelectorModel)\n",
+    "ui_final"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Enforcing the constraint between polynomial order and window size\n",
+    "\n",
+    "There are a few options here:\n",
+    "\n",
+    "1. Change one or both of the ranges of `widnow_size` and `polynomial_order`\n",
+    "    1. When window size is being changed and a conflict arises, either\n",
+    "        1. update the maximum value of the polynomial order to be consistent with window size, with the side effect that the value will be updated to.\n",
+    "        1. update the value of `polynomial_order` but not its maximum.\n",
+    "    1. When the polynomial order is changed, either\n",
+    "        1. update the minimum allowed window size, or\n",
+    "        1. update the value of the window size.\n",
+    "1. reset the controls to a valid value and display a warning to the user.\n",
+    "\n",
+    "It does not matter so much which of these you choose, but you do need to choose one."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the first version of the dashboard we did choice 1.A.a -- when `window_size` changed the `polynomial_order` was set to the smaller of 10 and `window_size - 1`.\n",
+    "\n",
+    "Here we will implement 2 because it is a little different than what we did in the first round of the dashboard. It also places the responsibility of deciding the correct course of action (increase `window_size` or decrease `polynomial-order`) on the user. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our approach is to observe the `_value` of the widget, try to make a valid model out of it, and if that fails set the widget to the old value. `ipyautoui` will take care of displaying an appropriate error message for us.\n",
+    "\n",
+    "The observe below is created using a [*closure*](https://en.wikipedia.org/wiki/Closure_(computer_programming)), which is a function that is created inside of another function where the inner function uses some of the variables form the outer function. Here we do that instead of treating `ui` as a global variable."
```

Comment:
Typo
```suggestion
   "The observe below is created using a [*closure*](https://en.wikipedia.org/wiki/Closure_(computer_programming)), which is a function that is created inside of another function where the inner function uses some of the variables from the outer function. Here we do that instead of treating `ui` as a global variable."
```